### ISSUE TITLE
Unique attribute in a repeatable component field leads save error when insert more component entries.

### ISSUE THREAD

**Issue by smoothdvd** [strapi/strapi#12541]
### Describe the bug

If a repeatable component has a field, which has unique attribute, insert this component multiple times will lead failure on unique validation.

### Steps to reproduce the behavior

1. Create a repeatable component name "foo"
2. Create a field named "title" in foo component, and enable "unique" validation on it
3. Add component foo to a content-type TEST
4. Add first entry with component  foo ( title => "test1") and save entry
5. Add second entry with component foo ( title => "test2" and save entry
6. See error: Failed to load resource: the server responded with a status of 400 (Bad Request)

```
{
    "data": null,
    "error": {
        "status": 400,
        "name": "ValidationError",
        "message": "14 errors occurred",
        "details": {
            "errors": [
                {
                    "path": [
                        "categories",
                        "0",
                        "name"
                    ],
                    "message": "This attribute must be unique",
                    "name": "ValidationError"
                },
...
```

### Expected behavior

Correct save multiple repeatable component which has contain unique validation field.

### Code snippets

If applicable, add code samples to help explain your problem.

### System

- Node.js version: 14.x
- NPM version: 6.14.15
- Strapi version: 4.0.7
- Database: PostgreSQL 13
- Operating system: macOS 12.2.1

**Comment by the94air** [comment]
I am running into the same issue with the same procedures as @mmarseu. For now the only way around is to disable the field unique option.

**Comment by mmarseu** [comment]
@gu-stav The restaurant example was only for illustration. My real use case is a different one. I was evaluating strapi to use as a backend for a product information site we'll be building in future. Take that into account, when reading my description below: _I don't actually know strapi very well and might have used it in unintended or inefficient ways._

We have a product portfolio with lots of _models_ belonging to different _classes_ and different _series_. That concept seems to map rather well to strapi collections. 
However, each model can also come in different _variants_. The variants of a model share a lot of common information but also have some pieces of information which are different. Among those is the _part number_ (or order number, if you like), which is unique.

My idea was to have _model_ be a collection type and _variant_ be a repeatable component on each _model_. That way, I can put common information in the fields of _model_ and the rest into the component. _Part number_ would have been part of the component and would have been unique.
Obviously, I would still have liked to be able to edit common information on the _model_, even if there are already any _variants_.

In more generic terms, personally, I would love to have the choice to have unique fields inside components be either:
- unqiue within the context of the parent content-type
- globally unique.
But the latter at least might be out of scope for components in strapi. As I said, I might be modelling things wrongly due to my lack of experience.

**Comment by tommyboylab** [comment]
I'm having the same issue as the OP. 

My case is I use a repeatable component to hold my navigationLink items which I set as a Single Type that I fetch dynamically with the page load. 

 I use a repeatable component as all the items are the same (a Name and a URL attribute) with the only difference being a different `Name` and `URL`. Because of this, I would like to set the attributes to be unique so that I prevent accidentally adding the same URL to a different Name link.

**Comment by jirbjirbjirb** [comment]
We are experiencing the same issue as OP in our project, we have a repeatable component with a 'slug' input that needs to be unique, but we are unable to save a collection with the repeatable component included, as returns the message:  `Error: This attribute must be unique`

I have disabled the unique field for the moment, but hope this is fixed soon

**Comment by Kraloz** [comment]
any workaround?

**Comment by jirbjirbjirb** [comment]
@Kraloz nope, we are still just not using a unique field attribute

**Comment by dpodz** [comment]
Also having this issue. Hope we get an update soon.

> What is the context in which you would expect it to be unique?

I'd expect it to be unique within the context of the parent. I think you can't go wrong going with stricter = better, but any sense of local uniqueness would be great.

**Comment by SzymonOwczarek** [comment]
Similar issue here:

1. Create a single type 
2. Create a repeatable component with some unique number (small int) field
3. Input multiple components with a unique number in each unique number field in the component
4. Save the draft
5. Publish the draft gives back an error saying that the field must be unique even though every number is unique.

For now, un-checking the required field is a workaround, but it would be great to have unique fields in repeatable components working in Strapi Single Types.

**Comment by gu-stav** [comment]
@SzymonOwczarek We are still collecting use-cases to understand what a possible solution could look like. Do you mind describing what you are using the uid in components for and what your expectation would be in terms of how it works (maybe even combined with i18n)?

**Comment by jeffreytung7** [comment]
@gu-stav I am also looking for this feature. I have a collection of products, and it has a repeatable component called productSellers. Here you can add the name of the seller (for example, Amazon), the URL of the product, the price of the product, and the unique ID of the product provided by Amazon. 

It's repeatable because I'd like to add other sellers (for example, Walmart). The IDs need to be unique so that in the future, I don't accidentally create two products in my collection that points to the same product on Amazon/Walmart. 

In other words having the UID would work to prevent me from creating duplicate products.

**Comment by manikanta47052** [comment]
is this resolved? im facing this issue.

**Comment by manikanta47052** [comment]
@gu-stav can you please let me know if there is any solution implemented for this issue. Using UID field is must for my use case.

**Comment by gu-stav** [comment]
@manikanta47052 Do you mind sharing your use-case with us? Right now we do not have a solution in place and we haven't prioritized this issue yet.

**Comment by manikanta47052** [comment]
@gu-stav I have exact use case like mentioned above. If I add unique attribute to a field in repeatable component, when im trying to add different content to that unique field getting 400 error.

**Comment by PhungVanHoa** [comment]
Any news ?

**Comment by spark-spartan** [comment]
@gu-stav I am seeing this too. I have a few use cases centred around the same idea which is to serve Questions in an Assessment style, where we have a ranked list as the answer. I.e. place a floppy disk, cd, dvd and blu-ray in invention time order. We are trying to achieve this in the following way:

- Question collection (basic, just question)
- Component Answers (answer text, rank numeric)
- within the rank numeric I want to set a min and max, 1-5 as each Q will have 5 answers
- and the unique field property so each rank position is only used once

Ideally we want to keep the rank position tied to the question and the ranks aren't reusable collections

**Comment by spark-spartan** [comment]
happy to be of some use!

For my use case it would be Entity uniqueness I need

Yeah makes perfect sense. Is the idea for Content Type and Entities that the IDs would remain unique across the whole component? Rather than also making IDs unique at Content Type or Entity?

**Comment by mhle** [comment]
I'm having a similar issue, error on save when I have multiple components that have unique constraint on a text field, even though the values are unique have not been used before. Any plans to fix this?

**Comment by Jackmekiss** [comment]
Same here

**Comment by Kartik4152** [comment]
Getting same error while trying to use the uuid plugin inside components in dynamic zone.  Any updates on this?

**Comment by derrickmehaffy** [comment]
Most likely this will not be addressed in Strapi v4 and will be handled in Strapi v5 as in order to properly fix this would require at least 1 breaking change (possibly more).

I'm going to lock this issue for now as I believe we have all the information we need.

**Comment by MarionLemaire** [comment]
Should be fixed in this week's v5 beta release, let's check after Wednesday 19th

**Comment by MarionLemaire** [comment]
Tested on v5.0.0-beta.13 : works for me, but date field uniqueness validation is not simultaneous with all other fields (mail, number, text) uniqueness validation. First you get a warning for uniqueness of text, number and mail fields, then you validate and you get warning for date field. 
(+ there is some random issue on date field and date picker, so let's be careful with date fields uniqueness as the date might be changed at validation)

**Comment by jhoward1994** [comment]
This issue has been fixed in v5 from 5.0.0-beta.13 onwards via https://github.com/strapi/strapi/pull/20427

The simultaneous validation issue will be tracked separately
